### DISCUSSION TITLE
Suggestion For Coding Guidelines

### DISCUSSION THREAD

**Discussion by RohitSrinivasRG** [Mindgrove-Technologies/mindgrove-bsc#11] [General]
This Discussion contains all the discussion regarding the decision behind bsv coding guidelines

**Comment by RohitSrinivasRG** [comment]
# Update register only at the end of the rule.
@Mindgrove-Technologies/rtl 

As the title says, we write the rule such that the register updates  only happen at the end of the rule. 

This helps us to understand what register are updated by the rule in a glance. One problem i constantly face in qspi is the registers are getting updated in many rules. The the rules are not particularly small, and the register are getting update in many if else clauses making it difficult to understand the flow.  

Pros:
- Understand the registers affected by the rule immediately when going through a code
- Easy to debug in case of any problem

Cons:
- Could be more prone to unintentional update of value multiple time in the same cycle leading to incorrect behaviour

**Comment by shash** [comment]
This works if we restrict the length of a rule to a certain amount - say 20-30 lines (which is good practice anyway). Anything further should be split off into functions / methods so that the rule itself is readable.

**Reply by mounakrishna99** [reply]
No even if the rule is large. It is better to have register updates at the end of the rule.
Having register updates in the middle of rules, can bring unnecessary double write errors by mistake.

Splitting big rules as functions/methods for readablity is necessary as you said.
